cpp/test/IceSSL/configuration/OpenSSLTests.cpp

```diff
@@ -117,6 +114,84 @@ clientValidatesServerUsingCAFile(Test::TestHelper* helper, const string& testDir
     cout << "ok" << endl;
 }
 
+void
+clientValidatesServerUsingValidationCallback(Test::TestHelper* helper, const string& testDir)
+{
+    cout << "client validates server certificate using validation callback... " << flush;
+
+    const string serverCertFile = testDir + "/../certs/s_rsa_ca1_pub.pem";
+    const string serverKeyFile = testDir + "/../certs/s_rsa_ca1_priv.pem";
+    SSL_CTX* serverSSLContext = SSL_CTX_new(TLS_server_method());
+    SSL_CTX_use_certificate_chain_file(serverSSLContext, serverCertFile.c_str());
+    SSL_CTX_use_PrivateKey_file(serverSSLContext, serverKeyFile.c_str(), SSL_FILETYPE_PEM);
+    SSL_CTX_set_default_passwd_cb(serverSSLContext, passwordCallback);
+
+    // The server certificate is not trusted by the client CA, but the validation callback accepts the server
+    // certificate.
+    const string clientCAFile = testDir + "/../certs/cacert2.pem";
+    SSL_CTX* clientSSLContext = SSL_CTX_new(TLS_client_method());
+    SSL_CTX_load_verify_file(clientSSLContext, clientCAFile.c_str());
+
+    try
+    {
+        Ice::SSL::ServerAuthenticationOptions serverAuthenticationOptions{
+            .serverSSLContextSelectionCallback = [serverSSLContext](const string&)
+            {
+                SSL_CTX_up_ref(serverSSLContext);
+                return serverSSLContext;
+            }};
+        Ice::CommunicatorHolder serverCommunicator(createServer(serverAuthenticationOptions, helper));
+
+        Ice::SSL::ClientAuthenticationOptions clientAuthenticationOptions{
+            .clientSSLContextSelectionCallback =
+                [clientSSLContext](const string&)
+            {
+                SSL_CTX_up_ref(clientSSLContext);
+                return clientSSLContext;
+            },
+            .serverCertificateValidationCallback = [](bool, X509_STORE_CTX*, const Ice::SSL::ConnectionInfoPtr&)
+            { return true; }};
+        Ice::CommunicatorHolder clientCommunicator(createClient(clientAuthenticationOptions));
+        ServerPrx obj(clientCommunicator.communicator(), "server:" + helper->getTestEndpoint(20, "ssl"));
+        obj->ice_ping();
+    }
+    catch (...)
+    {
+        SSL_CTX_free(serverSSLContext);
+        SSL_CTX_free(clientSSLContext);
+        throw;
+    }
+    SSL_CTX_free(serverSSLContext);
+    SSL_CTX_free(clientSSLContext);
+    cout << "ok" << endl;
+}
+
+void
+clientValidatesServerUsingSystemTrustedRootCertificates(Test::TestHelper*, const string&)
+{
+    cout << "client validates server using system trusted root certificates... " << flush;
+    SSL_CTX* clientSSLContext = SSL_CTX_new(TLS_client_method());
+    SSL_CTX_set_default_verify_paths(clientSSLContext);
+    Ice::SSL::ClientAuthenticationOptions clientAuthenticationOptions{
+        .sslNewSessionCallback = [](::SSL* ssl, const string& host)
+        {
+            // Enable SNI for connecting to the Glacier2 router behind NGINX proxy.
```

Comment:
SNI is needed here also for CloudFront, which is the first stop and requires SNI. 

I thought we performed SNI by default, is that not the case?

Reply:
> I thought we performed SNI by default, is that not the case?

Not enabled by default, Schannel enables it by default, but with OpenSSL and SecureTransport you have to use their respective APIs.

Keep in mind that with this API we provide mostly not defaults, and neither OpenSSL or SecureTransport enable it by default.

Reply:
I think this is something we should consider enabling by default.

Reply:
We enable SNI by default in 3.7 correct?

Reply:
We do not, see https://doc.zeroc.com/ice/3.7/property-reference/icessl#id-.IceSSL.*v3.7-IceSSL.CheckCertName